### PR TITLE
compilable version

### PR DESCRIPTION
**Pull request by achernigin1987**

### CODE REVIEW

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 1]
```diff
@@ -63,30 +60,30 @@ namespace Baikal
             }
 
             return std::make_unique<InferenceImpl>(model_path,
-                                                   gpu_memory_fraction,
-                                                   visible_devices,
-                                                   width,
-                                                   height,
+                                                   params.gpu_memory_fraction,
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: I'd suggest to rename `MLDenoiserParams` to `InferenceParams` and pass them to InferenceImpl constructor.

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 2]
```diff
@@ -63,30 +60,30 @@ namespace Baikal
...
+                                                   params.visible_devices,
+                                                   params.width,
+                                                   params.height,
                                                    input_channels);
         }
 
-        MLDenoiser::MLDenoiser(const CLWContext& context, Inference::Ptr inference, MLDenoiserInputs inputs)
-                   : m_inference(std::move(inference))
+        MLDenoiser::MLDenoiser(CLWContext context, const MLDenoiserParams& params)
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: Passing by value here kills the idea of forward declaration. It was a comment on Alexander's code, and now it is back.

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 3]
```diff
@@ -266,7 +249,7 @@ namespace Baikal
 
             if (!clw_inference_output)
             {
-                throw std::runtime_error("MLDenoiser::Apply(...): can not cast output");
+                throw std::runtime_error("Denoise::Apply(...): can not cast output");
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: The class name is not `Denoise`.

File: Baikal/PostEffects/ML/denoiser.h [hunk 4]
```diff
@@ -46,17 +48,19 @@ namespace Baikal
...
 
-        std::unique_ptr<Inference> CreateMLDenoiser(MLDenoiserInputs inputs,
-                                                    float gpu_memory_fraction,
-                                                    std::string const& visible_devices,
-                                                    std::size_t width,
-                                                    std::size_t height);
+        struct MLDenoiserParams {
+            MLDenoiserInputs inputs;
+            float gpu_memory_fraction;
+            std::string const& visible_devices;
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.h]: `std::string` to avoid dangling references.

File: Baikal/RenderFactory/render_factory.h [hunk 5]
```diff
@@ -26,12 +26,25 @@
...
 {
     class Renderer;
     class Output;
     class PostEffect;
-    
+
+#ifdef ENABLE_MLDENOISER
+    namespace PostEffects
+    {
+        class MLDenoiserParams;
```
Comment (dtarakanov1) [inline Baikal/RenderFactory/render_factory.h]: `struct`

File: Baikal/RenderFactory/render_factory.h [hunk 6]
```diff
@@ -26,12 +26,25 @@
...
 
+
 namespace Baikal
 {
     class Renderer;
     class Output;
     class PostEffect;
-    
+
+#ifdef ENABLE_MLDENOISER
```
Comment (dtarakanov1) [inline Baikal/RenderFactory/render_factory.h]: There are too many `#ifdef`s. I propose to keep a single `#ifdef` inside the factory method (throw if the macro is undefined) and make the rest of definitions always visible.

File: Baikal/RenderFactory/render_factory.h [hunk 7]
```diff
@@ -26,12 +26,25 @@
...
+    namespace PostEffects
+    {
+        class MLDenoiserParams;
+    }
+
+    enum class MLPostEffectType
```
Comment (dtarakanov1) [inline Baikal/RenderFactory/render_factory.h]: Why `MLPostEffectType`, not just `PostEffectType`?

File: BaikalStandalone/Application/cl_render.cpp [hunk 8]
```diff
@@ -50,13 +50,19 @@ namespace Baikal
     AppClRender::AppClRender(AppSettings& settings, GLuint tex) : m_tex(tex), m_output_type(Renderer::OutputType::kColor)
     {
         InitCl(settings, m_tex);
-        LoadScene(settings);
 
-#ifdef ENABLE_ML_DENOISER
+#if defined(ENABLE_DENOISER)
+        m_clw_denoiser = std::make_unique<ClwDenoiseController>(
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.cpp]: For some reason the denoisers are very different by API.

File: BaikalStandalone/Application/cl_render.cpp [hunk 9]
```diff
@@ -50,13 +50,19 @@ namespace Baikal
...
+                &(m_cfgs[m_primary]),
+                RenderFactory<Baikal::ClwScene>::PostEffectType::kWaveletDenoiser, m_width, m_height);
+        m_clw_denoiser->Init();
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.cpp]: `Init` is evil.

File: BaikalStandalone/Application/cl_render.cpp [hunk 10]
```diff
@@ -397,32 +398,22 @@ namespace Baikal
...
-        auto radius = 10U - RadeonRays::clamp((sample_cnt / 16), 1U, 9U);
-        auto position_sensitivity = 5.f + 10.f * (radius / 10.f);
-
-        const bool is_bilateral_denoiser = dynamic_cast<BilateralDenoiser*>(m_outputs[m_primary].denoiser.get()) != nullptr;
-
-        if (is_bilateral_denoiser)
+        if (m_clw_denoiser->GetType() == RenderFactory<ClwScene>::PostEffectType::kBilateralDenoiser)
         {
-
+            auto radius = 10U - RadeonRays::clamp((sample_cnt / 16), 1U, 9U);
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.cpp]: This code should be inside the denoiser class.

File: BaikalStandalone/Application/cl_render.cpp [hunk 11]
```diff
@@ -676,60 +667,35 @@ namespace Baikal
 #ifdef ENABLE_DENOISER
     void AppClRender::SetDenoiserFloatParam(const std::string& name, const float4& value)
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.cpp]: The method does not seem to be useful anymore.

File: BaikalStandalone/Denoise/denoise_controller.h [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+#include <memory>
+#include <unordered_map>
+
+namespace Baikal
+{
+    class DenoiseController {
+    public:
+        DenoiseController(
+                Config* config,
+                RenderFactory<ClwScene>::PostEffectType denoiserType,
```
Comment (dtarakanov1) [inline BaikalStandalone/Denoise/denoise_controller.h]: `denoiser_type`

File: BaikalStandalone/Denoise/denoise_controller.h [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+#include <memory>
+#include <unordered_map>
+
+namespace Baikal
+{
+    class DenoiseController {
+    public:
+        DenoiseController(
+                Config* config,
```
Comment (dtarakanov1) [inline BaikalStandalone/Denoise/denoise_controller.h]: `Config const&`?

File: BaikalStandalone/Denoise/denoise_controller.h [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+    protected:
+        using OutputPtr = std::unique_ptr<Baikal::Output>;
+
+        virtual void CreateRendererOutputs() = 0;
+
+        void AddRendererOutput(Renderer::OutputType type, bool add_to_input);
+
+        Config* m_config;
+        RenderFactory<ClwScene>::PostEffectType m_denoiserType;
```
Comment (dtarakanov1) [inline BaikalStandalone/Denoise/denoise_controller.h]: `m_denoiser_type`

File: Baikal/RenderFactory/render_factory.h [hunk 15]
```diff
@@ -66,6 +79,12 @@ namespace Baikal
         virtual 
         std::unique_ptr<PostEffect> CreatePostEffect(PostEffectType type) const = 0;
 
+#ifdef ENABLE_MLDENOISER
+        virtual
+        std::unique_ptr<PostEffect> CreateMLPostEffect(
```
Comment (dtarakanov1) [inline Baikal/RenderFactory/render_factory.h]: I thought the ML post effects will share code with non-ML effects.

File: BaikalStandalone/CMakeLists.txt [hunk 16]
```diff
@@ -45,7 +45,11 @@ set(SOURCES
     ${IMGUI_SORUCES}
     ${MAIN_SOURCES}
     ${UTILS_SOURCES}
-    ${KERNEL_SOURCES})
+    ${KERNEL_SOURCES}
+        Denoise/denoise_controller.h
```
Comment (dtarakanov1) [inline BaikalStandalone/CMakeLists.txt]: Indents.

File: Baikal/PostEffects/post_effect.h [hunk 17]
```diff
@@ -48,10 +49,15 @@ namespace Baikal
         // Data type to pass all necessary content into the post effect. 
         using InputSet = std::map<Renderer::OutputType, Output*>;
```
Comment (dtarakanov1) [inline Baikal/PostEffects/post_effect.h]: unordered_map?

File: BaikalStandalone/Application/cl_render.cpp [hunk 18]
```diff
@@ -151,13 +145,54 @@ namespace Baikal
...
+    {
+        Output* raw_output = nullptr;
+        auto it = m_renderer_outputs[device_idx].find(type);
+        if (it == m_renderer_outputs[device_idx].end())
+        {
+            const auto& config = m_cfgs.at(device_idx);
+            auto output = config.factory->CreateOutput(m_width, m_height);
+            raw_output = output.get();
+            config.renderer->SetOutput(type, raw_output);
+            config.renderer->Clear(RadeonRays::float3(0, 0, 0), *output);
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.cpp]: `float3()`

File: BaikalStandalone/Application/cl_render.cpp [hunk 19]
```diff
@@ -151,13 +145,54 @@ namespace Baikal
...
+
+            m_renderer_outputs[device_idx].emplace(type, std::move(output));
+        }
+        else
+        {
+            raw_output = it->second.get();
+        }
+
+        return raw_output;
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.cpp]: Let's use GetRendererOutput instead.

File: BaikalStandalone/Application/cl_render.cpp [hunk 20]
```diff
@@ -254,21 +289,18 @@ namespace Baikal
...
     {
-
         for (std::size_t i = 0; i < m_cfgs.size(); ++i)
         {
             if (i == static_cast<std::size_t>(m_primary))
             {
                 m_cfgs[i].controller->CompileScene(m_scene);
-                m_cfgs[i].renderer->Clear(float3(0, 0, 0), *m_outputs[i].output);
+                m_cfgs[i].renderer->Clear(
+                        float3(0, 0, 0),
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.cpp]: `float3()`

File: BaikalStandalone/Application/cl_render.cpp [hunk 21]
```diff
@@ -151,13 +145,54 @@ namespace Baikal
...
+    }
+
+    void AppClRender::GetOutputData(size_t device_idx, Renderer::OutputType type, RadeonRays::float3* data) const
+    {
+        m_renderer_outputs.at(device_idx).at(type)->GetData(data);
+    }
+
+    void AppClRender::AddPostEffect(size_t device_idx, RenderFactory<Baikal::ClwScene>::PostEffectType type)
+    {
+        m_post_effect = std::make_unique<PostEffectController>(&(m_cfgs[m_primary]), type, m_width, m_height);
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.cpp]: `&m_cfgs[m_primary]`

File: BaikalStandalone/Application/cl_render.cpp [hunk 22]
```diff
@@ -151,13 +145,54 @@ namespace Baikal
...
+
+        for (const auto& required_input : m_post_effect->GetInputTypes())
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.cpp]: `auto required_input`

File: BaikalStandalone/Application/cl_render.cpp [hunk 23]
```diff
@@ -674,62 +696,41 @@ namespace Baikal
     }
 
 #ifdef ENABLE_DENOISER
-    void AppClRender::SetDenoiserFloatParam(const std::string& name, const float4& value)
-    {
-        m_outputs[m_primary].denoiser->SetParameter(name, value);
+    RenderFactory<ClwScene>::PostEffectType AppClRender::GetDenoiserType() const{
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.cpp]: { from next line.

File: BaikalStandalone/Application/cl_render.h [hunk 24]
```diff
@@ -105,19 +90,24 @@ namespace Baikal
...
         void SetDenoiserFloatParam(const std::string& name, const float4& value);
         float4 GetDenoiserFloatParam(const std::string& name);
-        void CreateDenoiserOutputs(std::size_t cfg_index, int width, int height);
-        void SetDenoiserOutputs(std::size_t cfg_index) const;
-        void ClearDenoiserOutputs(std::size_t cfg_index) const;
         void RestoreDenoiserOutput(std::size_t cfg_index, Renderer::OutputType type) const;
 #endif
+
     private:
+        using RendererOutputs = std::map<Renderer::OutputType, std::unique_ptr<Output>>;
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.h]: unordered_map

File: BaikalStandalone/Denoise/post_effect_controller.cpp [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+#include "PostEffects/wavelet_denoiser.h"
+#include "PostEffects/ML/denoiser.h"
+
+
+namespace Baikal
+{
+    PostEffectController::PostEffectController(
+            Config* config,
+            RenderFactory<ClwScene>::PostEffectType type,
+            size_t width,
```
Comment (dtarakanov1) [inline BaikalStandalone/Denoise/post_effect_controller.cpp]: std::size_t

File: BaikalStandalone/Denoise/post_effect_controller.cpp [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        return m_post_effect->GetInputTypes();
+    }
+
+    void PostEffectController::SetParameter(std::string const& name, RadeonRays::float4 const& value)
+    {
+        m_post_effect->SetParameter(name, value);
+    }
+
+    RadeonRays::float4 PostEffectController::GetParameter(const std::string& name)
```
Comment (dtarakanov1) [inline BaikalStandalone/Denoise/post_effect_controller.cpp]: `std::string const&`

File: BaikalStandalone/Denoise/post_effect_controller.h [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+
+        PostEffect::InputTypes GetInputTypes() const;
+
+        RadeonRays::float4 GetParameter(const std::string& name);
+
+        void SetParameter(std::string const& name, RadeonRays::float4 const& value);
+
+        void SetCamera(Camera::Ptr camera);
+
+        void Process(PostEffect::InputSet const& input_set) const;
```
Comment (dtarakanov1) [inline BaikalStandalone/Denoise/post_effect_controller.h]: `Apply`, like the PostEffect method?

File: BaikalStandalone/Denoise/post_effect_controller.h [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+
+        Output* GetProcessedOutput() const;
```
Comment (dtarakanov1) [inline BaikalStandalone/Denoise/post_effect_controller.h]: GetOutput?

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 29]
```diff
@@ -41,11 +42,12 @@ namespace Baikal
         using float3 =  RadeonRays::float3;
         using OutputType = Renderer::OutputType;
 
-        std::unique_ptr<Inference> CreateMLDenoiser(MLDenoiserInputs inputs,
-                                                    float gpu_memory_fraction,
-                                                    std::string const& visible_devices,
-                                                    std::size_t width,
-                                                    std::size_t height)
+        std::unique_ptr<Inference> CreateDenoiserInference(
```
[the remark this answers is not in the document]
Comment (achernigin1987) [reply Baikal/PostEffects/ML/denoiser.cpp]: ok

File: Baikal/PostEffects/ML/denoiser.h [hunk 30]
```diff
@@ -22,15 +22,17 @@ THE SOFTWARE.
 
 #pragma once
 
-#include "PostEffects/post_effect.h"
 #include "PostEffects/ML/inference.h"
+#include "PostEffects/post_effect.h"
 #include "tensor.h"
 
+#include "CLW.h"
```
[the remark this answers is not in the document]
Comment (achernigin1987) [reply Baikal/PostEffects/ML/denoiser.h]: ok

File: BaikalStandalone/Application/cl_render.cpp (hunk 21 above)
Comment (achernigin1987) [reply BaikalStandalone/Application/cl_render.cpp]: ok

File: BaikalStandalone/Application/cl_render.cpp (hunk 22 above)
Comment (achernigin1987) [reply BaikalStandalone/Application/cl_render.cpp]: ok

File: BaikalStandalone/Application/cl_render.cpp (hunk 20 above)
Comment (achernigin1987) [reply BaikalStandalone/Application/cl_render.cpp]: ok

File: BaikalStandalone/Application/cl_render.cpp [hunk 31]
```diff
@@ -288,15 +333,20 @@ namespace Baikal
             if (std::atomic_compare_exchange_strong(&m_ctrl[i].newdata, &desired, 0))
             {
                 {
```
Comment (dtarakanov1) [inline BaikalStandalone/Application/cl_render.cpp]: Excessive.

**Pull request merged**